packages/core/src/modules/dcql/DcqlService.ts

```diff
@@ -0,0 +1,277 @@
+import type { AgentContext } from '../../agent'
+
+import { DcqlCredentialRepresentation, DcqlMdocRepresentation, DcqlQuery, DcqlSdJwtVcRepresentation } from 'dcql'
+import { injectable } from 'tsyringe'
+
+import { JsonValue } from '../../types'
+import { Mdoc, MdocApi, MdocDeviceResponse, MdocOpenId4VpSessionTranscriptOptions, MdocRecord } from '../mdoc'
+import { IPresentationFrame, SdJwtVcApi, SdJwtVcRecord } from '../sd-jwt-vc'
+import {
+  ClaimFormat,
+  W3cCredentialRecord,
+  W3cCredentialRepository,
+  W3cJsonLdVerifiablePresentation,
+  W3cJwtVerifiablePresentation,
+} from '../vc'
+
+import { DcqlError } from './DcqlError'
+import { DcqlQueryResult, DcqlCredentialsForRequest, DcqlPresentationRecord } from './models'
+import { dcqlGetPresentationsToCreate } from './utils'
+
+/**
+ * @todo create a public api for using dif presentation exchange
+ */
+@injectable()
+export class DcqlService {
+  /**
+   * Queries the wallet for credentials that match the given presentation definition. This only does an initial query based on the
+   * schema of the input descriptors. It does not do any further filtering based on the constraints in the input descriptors.
+   */
+  private async queryCredentialForPresentationDefinition(
+    agentContext: AgentContext,
+    dcqlQuery: DcqlQuery
+  ): Promise<Array<SdJwtVcRecord | W3cCredentialRecord | MdocRecord>> {
+    const w3cCredentialRepository = agentContext.dependencyManager.resolve(W3cCredentialRepository)
+
+    const formats = new Set(dcqlQuery.credentials.map((c) => c.format))
+    for (const format of formats) {
+      if (format !== 'vc+sd-jwt' && format !== 'jwt_vc_json' && format !== 'jwt_vc_json-ld' && format !== 'mso_mdoc') {
+        throw new DcqlError(`Unsupported credential format ${format}.`)
+      }
+    }
+
+    const allRecords: Array<SdJwtVcRecord | W3cCredentialRecord | MdocRecord> = []
+
+    // query the wallet ourselves first to avoid the need to query the pex library for all
+    // credentials for every proof request
+    const w3cCredentialRecords =
+      formats.has('jwt_vc_json') || formats.has('jwt_vc_json-ld')
+        ? await w3cCredentialRepository.getAll(agentContext)
+        : []
+    allRecords.push(...w3cCredentialRecords)
+
+    const sdJwtVcApi = this.getSdJwtVcApi(agentContext)
+    const sdJwtVcRecords = formats.has('vc+sd-jwt') ? await sdJwtVcApi.getAll() : []
+    allRecords.push(...sdJwtVcRecords)
+
+    const mdocApi = this.getMdocApi(agentContext)
+    const mdocRecords = formats.has('mso_mdoc') ? await mdocApi.getAll() : []
+    allRecords.push(...mdocRecords)
+
```

Comment:
The nice thing about dcql is that it is now much easier to preselect a subset of credentials based on doctype and vct.

Reply:
Yes but unfortunately it only works if all credentials for a specify type e.g sdjwtvc have a vct value requirement.